// @blueprintjs/datetime2/lib/cjs/classes.js
"use strict";
/*
 * Copyright 2023 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.dayPickerClassNameOverrides = exports.Classes = exports.DateRangePicker3Classes = exports.DatePicker3Classes = exports.DatePicker3CaptionClasses = exports.ReactDayPickerClasses = void 0;
var tslib_1 = require("tslib");
var classnames_1 = tslib_1.__importDefault(require("classnames"));
var core_1 = require("@blueprintjs/core");
var datetime_1 = require("@blueprintjs/datetime");
var RDP = "rdp";
var RDP_DAY = "".concat(RDP, "-day");
/** Class names applied by react-day-picker v8.x */
exports.ReactDayPickerClasses = {
    RDP: RDP,
    RDP_CAPTION: "".concat(RDP, "-caption"),
    RDP_CAPTION_DROPDOWNS: "".concat(RDP, "-caption_dropdowns"),
    RDP_CAPTION_LABEL: "".concat(RDP, "-caption_label"),
    RDP_DAY: RDP_DAY,
    RDP_DAY_DISABLED: "".concat(RDP_DAY, "_disabled"),
    RDP_DAY_HOVERED_RANGE: "".concat(RDP_DAY, "_hovered"),
    RDP_DAY_HOVERED_RANGE_END: "".concat(RDP_DAY, "_hovered_end"),
    RDP_DAY_HOVERED_RANGE_START: "".concat(RDP_DAY, "_hovered_start"),
    RDP_DAY_OUTSIDE: "".concat(RDP_DAY, "_outside"),
    RDP_DAY_RANGE_END: "".concat(RDP_DAY, "_range_end"),
    RDP_DAY_RANGE_MIDDLE: "".concat(RDP_DAY, "_range_middle"),
    RDP_DAY_RANGE_START: "".concat(RDP_DAY, "_range_start"),
    RDP_DAY_SELECTED: "".concat(RDP_DAY, "_selected"),
    RDP_DAY_TODAY: "".concat(RDP_DAY, "_today"),
    RDP_MONTH: "".concat(RDP, "-month"),
    RDP_NAV: "".concat(RDP, "-nav"),
    RDP_TABLE: "".concat(RDP, "-table"),
    RDP_VHIDDEN: "".concat(RDP, "-vhidden"),
};
exports.DatePicker3CaptionClasses = {
    DATEPICKER3_CAPTION: datetime_1.Classes.DATEPICKER_CAPTION,
    DATEPICKER3_DROPDOWN_MONTH: datetime_1.Classes.DATEPICKER_MONTH_SELECT,
    DATEPICKER3_DROPDOWN_YEAR: datetime_1.Classes.DATEPICKER_YEAR_SELECT,
    DATEPICKER3_NAV_BUTTON: "".concat(datetime_1.Classes.DATEPICKER, "-nav-button"),
    DATEPICKER3_NAV_BUTTON_NEXT: "".concat(datetime_1.Classes.DATEPICKER, "-nav-button-next"),
    DATEPICKER3_NAV_BUTTON_PREVIOUS: "".concat(datetime_1.Classes.DATEPICKER, "-nav-button-previous"),
};
exports.DatePicker3Classes = {
    DATEPICKER3_DAY: RDP_DAY,
    DATEPICKER3_DAY_DISABLED: exports.ReactDayPickerClasses.RDP_DAY_DISABLED,
    DATEPICKER3_DAY_IS_TODAY: exports.ReactDayPickerClasses.RDP_DAY_TODAY,
    DATEPICKER3_DAY_OUTSIDE: exports.ReactDayPickerClasses.RDP_DAY_OUTSIDE,
    DATEPICKER3_DAY_SELECTED: exports.ReactDayPickerClasses.RDP_DAY_SELECTED,
    DATEPICKER3_HIGHLIGHT_CURRENT_DAY: "".concat(datetime_1.Classes.DATEPICKER, "-highlight-current-day"),
    DATEPICKER3_REVERSE_MONTH_AND_YEAR: "".concat(datetime_1.Classes.DATEPICKER, "-reverse-month-and-year"),
};
exports.DateRangePicker3Classes = {
    DATERANGEPICKER3_HOVERED_RANGE: exports.ReactDayPickerClasses.RDP_DAY_HOVERED_RANGE,
    DATERANGEPICKER3_HOVERED_RANGE_END: exports.ReactDayPickerClasses.RDP_DAY_HOVERED_RANGE_END,
    DATERANGEPICKER3_HOVERED_RANGE_START: exports.ReactDayPickerClasses.RDP_DAY_HOVERED_RANGE_START,
    DATERANGEPICKER3_REVERSE_MONTH_AND_YEAR: "".concat(datetime_1.Classes.DATERANGEPICKER, "-reverse-month-and-year"),
    DATERANGEPICKER3_SELECTED_RANGE_END: exports.ReactDayPickerClasses.RDP_DAY_RANGE_END,
    DATERANGEPICKER3_SELECTED_RANGE_MIDDLE: exports.ReactDayPickerClasses.RDP_DAY_RANGE_MIDDLE,
    DATERANGEPICKER3_SELECTED_RANGE_START: exports.ReactDayPickerClasses.RDP_DAY_RANGE_START,
    DATERANGEPICKER3_TIMEPICKERS_STACKED: "".concat(datetime_1.Classes.DATERANGEPICKER_TIMEPICKERS, "-stacked"),
};
/** Class names for next-gen @blueprintjs/datetime2 "V3" components */
exports.Classes = tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, datetime_1.Classes), exports.DatePicker3CaptionClasses), exports.DatePicker3Classes), exports.DateRangePicker3Classes);
/**
 * Class name overrides for components rendered by react-day-picker. These are helpful so that @blueprintjs/datetime2
 * can have more predictable and standard DOM selectors in custom styles & tests.
 */
exports.dayPickerClassNameOverrides = {
    /* eslint-disable camelcase */
    button: (0, classnames_1.default)(core_1.Classes.BUTTON, core_1.Classes.MINIMAL),
    // no need for button "reset" styles since the core Button styles handle that for us
    button_reset: undefined,
    dropdown_month: exports.DatePicker3CaptionClasses.DATEPICKER3_DROPDOWN_MONTH,
    dropdown_year: exports.DatePicker3CaptionClasses.DATEPICKER3_DROPDOWN_YEAR,
    nav_button: exports.DatePicker3CaptionClasses.DATEPICKER3_NAV_BUTTON,
    nav_button_next: exports.DatePicker3CaptionClasses.DATEPICKER3_NAV_BUTTON_NEXT,
    nav_button_previous: exports.DatePicker3CaptionClasses.DATEPICKER3_NAV_BUTTON_PREVIOUS,
    /* eslint-enable camelcase */
};
